src/fixit/lsp.py

```diff
@@ -0,0 +1,162 @@
+        """
+        Side-effect: publishes Fixit diagnostics to the LSP client.
+        """
+        generator = self.diagnostic_generator(uri)
+        if not generator:
+            return
+        diagnostics = []
+        for result in generator:
+            violation = result.violation
+            if not violation:
+                continue
+            diagnostic = Diagnostic(
+                Range(
+                    Position(  # LSP is 0-indexed; fixit line numbers are 1-indexed
+                        violation.range.start.line - 1, violation.range.start.column
+                    ),
+                    Position(violation.range.end.line - 1, violation.range.end.column),
+                ),
+                violation.message,
+                severity=DiagnosticSeverity.Warning,
+                code=violation.rule_name,
+                source="fixit",
+            )
+            diagnostics.append(diagnostic)
+        self.lsp.publish_diagnostics(uri, diagnostics, version=version)
+
+    def debounced_validator(self, uri: str) -> Callable[[int], None]:
+        """
+        Per-URI debounced validation function. See: LSP.validate
+        """
+        if uri not in self._debounced_validators:
+            self._debounced_validators[uri] = debounce(
+                self.lsp_options.debounce_interval
+            )(lambda version: self.validate(uri, version))
+        return self._debounced_validators[uri]
+
+    def set_handlers(self) -> None:
+        """
+        Side-effect: mutates self.lsp to make it respond to
+        `textDocument/didOpen`, `textDocument/didChange`, `textDocument/formatting`
+        """
+
+        @self.lsp.feature(TEXT_DOCUMENT_DID_OPEN)
+        def _(params: DidOpenTextDocumentParams):
+            self.debounced_validator(params.text_document.uri)(
+                params.text_document.version
+            )
+
+        @self.lsp.feature(TEXT_DOCUMENT_DID_CHANGE)
+        def _(params: DidChangeTextDocumentParams):
+            self.debounced_validator(params.text_document.uri)(
+                params.text_document.version
+            )
+
+        @self.lsp.feature(TEXT_DOCUMENT_FORMATTING)
+        def _(params: DocumentFormattingParams):
+            generator = self.diagnostic_generator(
+                params.text_document.uri, autofix=True
+            )
+            if generator is None:
+                return None
+
+            captured = capture(generator)
+            for _ in captured:
+                pass
+            formatted_content = captured.result
+            if not formatted_content:
+                return None
+
+            doc = self.lsp.workspace.get_document(params.text_document.uri)
+            entire_range = Range(
+                start=Position(line=0, character=0),
+                end=Position(line=len(doc.lines) - 1, character=len(doc.lines[-1])),
+            )
+
+            return [TextEdit(new_text=formatted_content.decode(), range=entire_range)]
```

Comment:
Would like this to be in a separate method the way `validate` is. Maybe just something like `format(...) -> List[TextEdit]`?